ntp.go

```diff
@@ -272,8 +366,8 @@ func getTime(host string, opt QueryOptions) (*msg, error) {
 }
 
 // TimeV returns the current time from the remote server host using the
-// requested version of the NTP protocol. The version may be 2, 3, or 4;
-// although 4 is most typically used.
+// requested version of the NTP protocol, returns local time on failure. The
+// version may be 2, 3, or 4; although 4 is most typically used.
```

Comment:
I suggest the following:

```go
// TimeV returns the current time from the remote server host using the
// requested version of the NTP protocol. On error, it returns the local
// time. The version may be 2, 3, or 4.
```

Reply:
@beevik, I would like to know your opinion. Should Time and TimeV check basic validity of the packet before returning the time?

Reply:
I'm conflicted.  On the one hand I'd like to keep `Time` and `TimeV` dirt simple, because most people use this package (particularly the `Time` APIs) as a quick and dirty way of getting a reasonably accurate reflection of the current time.  If they have to handle validity errors, it becomes less easy to rely on.  (It's also not clear what they should do in that circumstance.)  On the other hand, I don't like the possibility that `Time` will return an invalid time.

A couple questions for you: when the time reported by an NTP server is "invalid", what are the practical reasons for that?  Just *how* far off from actual time is it possible to be in a typical case?  Is this likely when using one of the standard pool NTP servers? I'm not an NTP expert by any means, so I'd like to get a better understanding of how likely an "invalid" time is.

Reply:
> It's also not clear what they should do in that circumstance

Retry using another NTP server if they have network connectivity.

Probably, the only thing the library can do is to retry it for the client if there is more than one IP address for passed domain name. Or query all addresses for the name 8 times and select "the best one" using proper [false-ticker detection](https://tools.ietf.org/html/rfc5905#appendix-A.5.5.1) (I'm unsure if I'm going to contribute the code for that). One of the API issues is that only the last `err` may be preserved in case of network errors, another one is that it may be good to expose that non-trivial algorithm to be used with set of IPs instead of single DNS name.

I think that discarding invalid packets is the way to go as `Validate()` is really just a hygienic check on the packet. If anyone needs more complex, issue (or PR) should be opened :-)

> the practical reasons for that

E.g. server reboot. I've seen servers in NTP pool reporting stratum=16 with some significant offset (seconds). These servers are eventually removed from rotation because there is some [monitoring system](http://www.pool.ntp.org/scores/139.59.19.184) that detects them, but client may eventually get some responses from broken time server.

>  Just how far off from actual time is it possible to be in a typical case

I don't know a typical case. That broken pool server that I've occasionally seen was several seconds away. That may be already unsuitable for purposes of TOTP (Google Authenticator).

Reply:
> reasonably accurate reflection of the **current** time

NB, it does not account for RTT, so it's quite wrong when RTT is high :)